executor/prepared.go

```diff
@@ -97,6 +97,10 @@ func NewPrepareExec(ctx sessionctx.Context, is infoschema.InfoSchema, sqlTxt str
 // Next implements the Executor Next interface.
 func (e *PrepareExec) Next(ctx context.Context, chk *chunk.Chunk) error {
 	vars := e.ctx.GetSessionVars()
+	vars.InPrepare = true
```

Comment:
It might be another clean-up issue. How about we treat it as another PR?

Reply:
It is up to your choice. I prefer this because the one line change can remove all `InPrepare` changes in this patch.

Reply:
@eurekaka I don't think so. Since the value of the parameter can be ``nil``, we couldn't know that we are in the in-prepare stage depending on the value.

Reply:
Why must we know that we are in the "in-prepare" stage as long as we set the parameters to null? why prepare should be special in this case?

Reply:
@eurekaka 
This is because ``plannercore.BuildLogicalPlan`` is called whenever ``prepare {stmt_id} from '{stmt}'`` statements are issued. Here, ``plannercore.BuildLogicalPlan`` executes ``{stmt}`` in the ``prepare`` context with no given parameter values. It is different to the case, ``execute {stmt_id}``.  Here, ``execute {stmt_id}`` can be issued with no given parameter values. The former should not care if the parameter values are given, but the latter should care. Thus, we should distinguish each case to handle each statement correctly.   

https://github.com/pingcap/tidb/blob/4bcdb6d8ee7906947008603d95c543cd978b1106/executor/prepared.go#L167

Reply:
> Here, plannercore.BuildLogicalPlan builds a logical plan for the {stmt} in the prepare context with no given parameter values.

All `ParamMarkerExpr` in `{stmt}` is filled with integer 0 before `BuildLogicalPlan`.
```
    // We try to build the real statement of preparedStmt.
    for i := range prepared.Params {
        prepared.Params[i].(*driver.ParamMarkerExpr).Datum = types.NewIntDatum(0)
    }
```

> The former should not care if the parameter values are given, but the latter should care.

They all do care in current implementation at least. `prepare` is not a special case for `BuildLogicalPlan`.

Reply:
> For the former case, we can set the default parameter value as nil (i.e.,prepared.Params[i].(*driver.ParamMarkerExpr).Datum.SetNull(). But it causes other side effects.

That's why I suggest setting default parameter value as empty string, because user variable is string type:
prepared.Params[i].(*driver.ParamMarkerExpr).Datum = **types.NewStringDatum("")**

Reply:
Got it! The updated PR should address the issue.

Reply:
@eurekaka 
Are the CI errors related with changing of the default parameter value?

Reply:
@eurekaka, Should we revert the latest change?

Reply:
It is related. You can revert the latest change then, I can find out why it fails using https://github.com/pingcap/tidb/pull/8535. Sorry for the misleading here.

Reply:
@eurekaka You're welcome.  I've reverted the latest changes.

Reply:
Thank you for the analysis!